pkg/proxy/proxy/transparent.go

```diff
@@ -117,60 +118,104 @@ func NewTransparentProxyService(p *goproxy.ProxyHttpServer, ca *tls.Certificate,
 
 // ProxyHTTP serves an endpoint that transparently redirects HTTP connections to the proxy server.
 // This endpoint also explicitly (i.e. non-transparently) proxies HTTP and TLS connections.
-func (t TransparentProxyService) ProxyHTTP(addr string) {
+func (t TransparentProxyService) ProxyHTTP(addr string) func(context.Context) error {
 	ln, err := net.Listen("tcp", addr)
 	if err != nil {
 		log.Fatalf("Error listening for http connections - %v", err)
 	}
-	if err := http.Serve(ln, t.Proxy); err != nil {
-		log.Fatalln(err)
+	server := &http.Server{
+		Addr:    addr,
+		Handler: t.Proxy,
 	}
+	go func() {
+		if err := server.Serve(ln); err != nil {
+			log.Println(err)
+		}
+	}()
+	return func(ctx context.Context) error { return server.Shutdown(ctx) }
 }
 
 // ProxyTLS serves an endpoint that transparently redirects TLS connections to the proxy server.
-func (t TransparentProxyService) ProxyTLS(addr string) {
+func (t TransparentProxyService) ProxyTLS(addr string) func(context.Context) error {
 	ln, err := net.Listen("tcp", addr)
 	if err != nil {
 		log.Fatalf("Error listening for https connections - %v", err)
 	}
-	for {
-		log.Println("Awaiting TLS connection")
-		c, err := ln.Accept()
-		if err != nil {
-			log.Printf("Error accepting new connection - %v", err)
-			continue
-		}
-		go func(c net.Conn) {
-			log.Println("Connection from ", c.RemoteAddr())
-			conn, hello, err := handshake.PeekClientHello(c)
-			if err != nil {
-				log.Printf("Error accepting new connection - %v", err)
+	var inflightRequests sync.WaitGroup
+	t.Proxy.OnResponse().DoFunc(func(resp *http.Response, ctx *goproxy.ProxyCtx) *http.Response {
+		defer inflightRequests.Done()
+		return resp
+	})
+	done := make(chan bool)
+	go func() {
+		for {
+			select {
+			case <-done:
 				return
+			default:
+				log.Println("Awaiting TLS connection")
+				c, err := ln.Accept()
+				if err != nil {
+					log.Printf("Error accepting new connection - %v", err)
+					continue
+				}
+				inflightRequests.Add(1)
+				go func(c net.Conn) {
+					log.Println("Connection from ", c.RemoteAddr())
+					conn, hello, err := handshake.PeekClientHello(c)
+					if err != nil {
+						log.Printf("Error accepting new connection - %v", err)
+						return
+					}
+					host := hello.ServerName
+					if host == "" {
+						log.Printf("Cannot support non-SNI enabled clients")
+						c.Close()
```

Comment:
inflightRequests won't get updated here or in the block above. I think you should add a `defer inflightRequests.Done()` at the top of the function to be sure it's resolved. (we'd need to make sure this code is respectful of some deadline/timeout to be entirely confident in the shutdown not hanging but I think it's fine as-is)

Reply:
Good point. I originally had `defer inflightRequests.Done()` at the top but realized this doesn't wait for the request to actually complete. Instead I added it in the `OnResponse` callback, but looking at the code goproxy actually returns a response before the request is complete. So I think a graceful shutdown is not possible unless we make upstream changes 🙃 

I just put `defer inflightRequests.Done()` for now so it doesn't hang if there's an error somewhere.